### DISCUSSION TITLE
Proposal: Grouping types via `using` aliases

### DISCUSSION THREAD

**Discussion by osamakawish** [dotnet/csharplang#5476] [Language Ideas]
## Summary

The goal is to group types into a single "grouped type" or as [#4628](https://github.com/dotnet/csharplang/discussions/4628) calls it, "intersection types" and name them via a using alias. These are to be implemented when defining variables or declaring inheritance:
```c#
using IContrastable = IEquatable and IComparable;

public Int32 : IContrastable<T> { ... }

public class C
{
    ...
    public IContrastable<T> contrastable;
    ...
}
```
## Motivation

Originally intended [specifically for class members and properties](https://github.com/dotnet/csharplang/discussions/5472), this acts as an alternative to `where` conditions in various situations, with the added benefits of allowing variables to have been implemented from various interfaces. That is, the goal was simply to let the compiler know that my class member should inherit from the respective type. 

Currently, I'm only proposing that we focus on grouping intersection types from [#4628](https://github.com/dotnet/csharplang/discussions/4628), as I believe they're more natural to the development of programming languages (ie. it's essentially like passing in multiple types to inherit from). Union types come with a series of issues that need to be addressed.

But as the discussion went on and the syntax developed, I realized the notation could be extended to existing inheritances involving where conditions, or repeated calls to inheritance from the same set of classes. Many classes already reuse the same sets of interfaces, particularly collections, which can often be grouped under a single "grouped type" (which I'll refer to Type Groups from here on out).

The syntax is straightforward, though there are two cases of this:

1. The **type group** has only interfaces:

   ```c#
   using IContrastable = IEquatable and IComparable;
   using IContrastable<T> = IEquatable<T> and IComparable<T>;
   ```

2. The **type group** also has a class:

   ```c#
   public interface I { ... }
   public abstract class A { ... }
   
   using X = I and A;
   ```

The use cases are for inheritances and variable/property declarations:

* Inheriting from a **type group**:

  ```c#
  public Int32 : IContrastable<T> { ... }
  public Y : X { ... }
  ```

* Having a class member or property with a **type group**:

  ```c#
  public IContrastable<T> contrastable;
  public IContrastable<T> Contrastable { get; set; }
  
  public X x;
  public X { get; set; }
  ```

  Though here, during variable declaration or property setting, the compiler tests if the object's class inherits from all the classes in the **type group**, rather than simply checking a single type that the variable is assigned to (which is currently the case). 

The use of the word `and` is deliberate. I think there are some conflicts that might arise in union classes which generally don't exist in intersection classes. However, if they get resolved, this notation will extend to union classes just as well. Nonetheless, so long as it doesn't people may mistakenly assume they're allowed to substitute the `and` with `or` which will lead to a dreaded compiler error. For now, it may be preferable to not use the `and` and treat is as default in the future when and if union classes are introduced, in order to minimize confusion:

```c#
public interface I { ... }
public abstract class A { ... }

using X = I A;
using IContrastable = IEquatable IComparable;
using IContrastable<T> = IEquatable<T> IComparable<T>;
```

As elaborated in [#5472](https://github.com/dotnet/csharplang/discussions/5472), this will lead to various improvements in code and refactoring, while maintaining readability.

Additionally, this code helps when part of the code is written by someone else, and also saves us from having to write interfaces or abstract classes that implement multiple interfaces.

### Syntax

First proposal so I don't really know how the notation for syntax work. If anyone wants to help me out, that would be great lol.

### Semantics

Again, not sure if I'm doing this right, but:

A **type group** is just a collection of types, containing 1 or more interfaces and 0 or 1 class.

* Any variable of the associated type group implements all its interfaces and its given class, if any.
* Variables can only be declared to type groups with a class (for obvious reasons).
* If a variable has an associated type group:
  * During declaration, the compiler should check if the value being assigned to the variable belongs to an object that is an instance of the type group's class, and that the value's class implements all classes in the type group.
  * All methods associated with the interfaces and the type group's class and interfaces are valid calls from the variable. It should behave identically to any other variable having a class that inherits from the type groups classes and interfaces.
* A type group cannot contain a sealed class. A sealed class cannot be inherited. All uses of type groups become redundant here. The inheritance is irrelevant, and variable declarations of sealed classes only make sense if the type group already inherits the sealed class. But at that point, the variable might as well have been declared in terms of the sealed class itself.
* Every class must inherit from either a class directly, or at most 1 type group with a class.
* Interfaces can only implement type groups that only contain interfaces.

All usages of type groups can be simplified into the cases of inheritance/implementation and variable declarations. 

## Drawbacks

* Performance maybe, though how much is unclear. I personally expect it to be slight in the case of variable declarations. As for inheritance, still relatively slight as far as I can think of.

Feel free to mention any that I haven't thought of though. 

## Alternatives

The `where` conditions are already an alternative in the case of inheritance. This feature adds the additional option to extend the idea of having the same behavior as where conditions into member declarations.

Another alternative is the idea of [intersection classes](https://github.com/dotnet/csharplang/discussions/4628) that was already mentioned as an alternative, but I believe that discussion may divert the subject towards also trying to implement union classes, which have some issues to be resolved.

## Unresolved Questions

None that I can think of. Even the unresolved questions and drawbacks raised over at [#4628](https://github.com/dotnet/csharplang/discussions/4628) only focus on union classes. Since this is geared particularly towards intersection classes at the moment, there's only a few general questions that need to be addressed with any other feature request, that I am unable to do so due to a lack of experience contributing and knowledge of back-end code.

* What other existing language features need to be considered when adding the feature?
* How complicated are the necessary changes that need to be made to the code-behind?
* How practical is the idea of implementing "type groups" as an additional feature of the language?

I'm also curious if anyone's got better suggestions for syntax, as this is just the best I could personally think of.

**Comment by HaloFour** [comment]
I don't like the idea of using aliases here, as these are not aliases.  You are describing some new kind of type, not referencing an existing one.

Moreso, this doesn't describe how these "aliases" or "intersections" actually work.  What is the actual type of a variable declared as one of these aliases?  It can't be the alias, because the alias doesn't exist.  It can't be _both_ of the aliased types, because a variable can't be declared as two different types.  The compiler might be able to get tricky there when it comes to locals and pretend that the intersection exists when the local is defined as some other concrete type and cast implicitly where necessary.  But the compiler can't do the same thing with properties or other exposed values, they need to be of exactly one type reference.

In short I think that this proposal runs into the same problems which has caused the other proposal to not go anywhere, specifically that without runtime support for a way to describe a type intersection _and_ to be able to treat that as a type reference that there really isn't a whole lot that can be done in this space.  Any compiler hackery will be a very leaky abstraction.  This is much more a problem of implementation than it is of syntax.

**Reply by osamakawish** [reply]
This is a question I've had then: Would it not be possible to do these intersections at compile-time?

Like during compile-time type declarations, instead of checking whether the type is identical to a requested type, you check whether the type of the object being given the assignment has all the interfaces and inherits the class of the target member/property.

```c#
public IContrastable<Int32> contrastable = 24;
```

The compiler should already know that:

* IContrastable is a mix of IComparable and IEquatable (as declared in OP)
* Int32 implements both

Therefore, should it not be able to use that information to allow the assignment? And similarly say, reject assignments to a string?

The `using alias` was me just providing syntax based on existing notation. Though that syntax can be changed into something that makes more sense.

**Reply by HaloFour** [reply]
The problem is exactly what type is `contrastable`?  It can't be `IContrastable<Int32>` as that type doesn't exist.  It can't be both `IEquatable<Int32>` and `IComparable<Int32>` as the field in the assembly can only be of one specific type.  So exactly what type is that field?

**Reply by osamakawish** [reply]
Okay. I understand the confusion. There isn't an actual type, in the sense that `typeof(TypeGroup)` wouldn't result in a clear value. And trying to access the type of the object via reflection wouldn't result in a proper value. 

I think I was thinking the type could be "anonymous" - in a weird sort of a way. Or at least the behaviour is similar in that it's not something known at compile time, and the runtime could maybe jumble up a name to assign it.

But it works a little differently than how types currently work. It's more of a collection of types which my previous comment suggests how the compiler could go about testing which declarations it allows and which it does not.

**Reply by osamakawish** [reply]
> The problem is exactly what type is `contrastable`?  It can't be `IContrastable<Int32>` as that type doesn't exist.  It can't be both `IEquatable<Int32>` and `IComparable<Int32>` as the field in the assembly can only be of one specific type.  So exactly what type is that field?

I mean, I agree it isn't defined by a type. But I guess the issue with that question is that it's like asking "what number is the set { 12, -7 }?". And I mean, it isn't a unique number, it's a group/set of numbers. And that question can become even more misleading when phrased as something like: "what is the zero of x^2 - 1?" - as it assumes that the zero must be unique. 

Similarly type groups aren't unique types, they're a collection of them. I just explained how I want these "Type Groups" to behave (as well as what features we could use for them), should they be added as a feature of the langauge.

I suppose that raises a myriad of questions regarding the issues and like that might arise from adding such a feature, and from extending the idea that "variables must have a unique type" to "variables must belong a unique type group".

**Reply by HaloFour** [reply]
But that's the problem.  Fields need to be a type.  Properties need to be a type.  Parameters need to be a type.  The runtime has no concept of "type collections" or "type groups", only types.  The C# compiler needs to emit an assembly that makes sense to the runtime and can be understood by other languages that run on .NET, and it needs to emit an assembly that has a stable ABI so that a recompilation doesn't break any assembly that depends on that code, and it needs to be explicitly defined in the language spec so that a recompilation by an alternate implementation of the compiler will emit the same results.

**Reply by osamakawish** [reply]
> But that's the problem.  Fields need to be a type.  Properties need to be a type.  Parameters need to be a type.  The runtime has no concept of "type collections" or "type groups", only types.  The C# compiler needs to emit an assembly that makes sense to the runtime and can be understood by other languages that run on .NET, and it needs to emit an assembly that has a stable ABI so that a recompilation doesn't break any assembly that depends on that code, and it needs to be explicitly defined in the language spec so that a recompilation by an alternate implementation of the compiler will emit the same results.

Fair enough. The implementation was where I needed advice for, since that's easily where I have the least knowledge.

But thanks for the criticism nonetheless. 

I suppose a feature like this would be more practical if we started an entirely new language with it though. But yeah, probably not easy for c#.

**Reply by osamakawish** [reply]
Wait. I mean, the point of proposals is to introduce new language features. My point is that can something that behaves like a type group not be added as feature in the language?

I just thought of a "workaround" which even the compiler may be able to implement: You use multiple variables with the same referenced object (if code below works, may need to cast to object first, but you get the gist of it):

```c#
IEquatable<int> eq = 24;
IComparable<int> comp = (IComparable<int>)eq;
```

It offers the same benefits, but VS won't suggest the methods of both under a single blanket variable, which is what I want. But here the benefit of having pre-casted members is what this adds to the language (which is basically what I'm going for with these "type groups").

The compiler/runtime doesn't need an inherent feature to be able to give a perception of that feature in the langauge.

**Reply by osamakawish** [reply]
Basically, if we do what I did in the previous comment, the type of the object will vary based on what value the variable/member/property has been assigned.

**Comment by YairHalberstadt** [comment]
This proposal seems to assume that https://github.com/dotnet/csharplang/discussions/4628 is implemented (or at least half of it), and will fall out naturally if it is from the existing rules about using statements. As such I'm not quite sure what it's adding?

**Reply by osamakawish** [reply]
Not saying it is implemented, but if it's implemented in the future, I was using that syntax.

**Reply by YairHalberstadt** [reply]
It seems to be just as hard as (half of) that proposal to implement, and fall directly out of that proposal, so I'm still not really seeing the point...

**Reply by osamakawish** [reply]
> It seems to be just as hard as (half of) that proposal to implement, and fall directly out of that proposal, so I'm still not really seeing the point...

Technically does fall out of the concept of intersection types. As once those are implemented, this is just like a using statement.

If you look over at the other proposal though, most questions and issues people raises were with regards to union types, not intersection types. It's why I deliberately want to focus on "intersection types"/"type groups" here.
